Sources/GRPCHTTP2Core/Compression/Zlib.swift

```diff
@@ -0,0 +1,423 @@
+/*
+ * Copyright 2024, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import CGRPCZlib
+import GRPCCore
+import NIOCore
+
+enum Zlib {
+  enum Method {
+    case deflate
+    case gzip
+
+    fileprivate var windowBits: Int32 {
+      switch self {
+      case .deflate:
+        return 15
+      case .gzip:
+        return 31
+      }
+    }
+  }
+}
+
+extension Zlib {
+  /// Creates a new compressor for the given compression format.
+  ///
+  /// This compressor is only suitable for compressing whole messages at a time. Callers
+  /// must ``initialize()`` the compressor before using it and must ``reset()`` it between
+  /// subsequent calls to ``compress(_:into:)``.
```

Comment:
Is there a reason why we couldn't simplify this API by calling `stream/deflateInit` in the initialiser, and doing the logic of `reset()` before calling `deflate` in `compress(_:into:)`? It would make it impossible to make mistakes (by forgetting to call these methods) while using the compressor.
Same question about the `Decompressor` below.

Reply:
Unfortunately not, initialising in the `init` results in a `Z_STREAM_ERROR` when you call compress (same goes for the decompress). The same is true if you use a static function to create and then initialize and return the initialized compressor/decompressor. I don't understand _why_ that's the case though. In v1 we _did_ do this in `init` but the compressor/decompressor were both classes but I'd rather save the allocation.

W.r.t. `reset()`, that seems like a reasonable idea. I'm not sure what the performance implications are but the docs say it does not free or reallocate internal state so _should_ be okay to do.

Reply:
Ah, that's unfortunate about the `init`.

W.r.t `reset()` - if the performance was to be affected by running it always in the `defer` block, we could instead add another flag (`isDirty` or something like that, signalling that we've already called `(de)compress`) and if that's set at the beginning of the `reset()` call, then we reset just before doing the rest of the logic. 
That way we'd be saving the cost of _one_ `reset()` (the one that would otherwise happen after the last (de)compression) - but well, I don't know if we care enough about this for now :D